From 1d08c8462c2004afe31ada86f0b1d7eac5e66f24 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Thu, 1 Aug 2013 17:04:08 +0900
Subject: [PATCH 1078/1083] drm: rcar-du: Disable alpha blending of lowest
 planes

The alpha blending operation is disabled in case of frame buffer planes.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_crtc.c  | 2 +-
 drivers/gpu/drm/rcar-du/rcar_du_plane.c | 8 +++++++-
 drivers/gpu/drm/rcar-du/rcar_du_plane.h | 2 ++
 drivers/gpu/drm/rcar-du/rcar_du_regs.h  | 1 +
 4 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
index 534443d..961725a 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
@@ -254,7 +254,7 @@ static void rcar_du_crtc_start(struct rcar_du_crtc *rcrtc)
 
 		if (plane->crtc != crtc || !plane->enabled)
 			continue;
-
+		plane->fb_plane = true;
 		rcar_du_plane_setup(plane);
 	}
 
diff --git a/drivers/gpu/drm/rcar-du/rcar_du_plane.c b/drivers/gpu/drm/rcar-du/rcar_du_plane.c
index 4e0c206..d080a8c 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_plane.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_plane.c
@@ -267,7 +267,12 @@ static void rcar_du_plane_setup_mode(struct rcar_du_plane *plane,
 		rcar_du_plane_write(rgrp, index, PnALPHAR,
 				    PnALPHAR_ABIT_X | plane->alpha);
 
-	pnmr = PnMR_BM_MD | plane->format->pnmr;
+	/* Disable alpha blending of lowest plane */
+	if (plane->fb_plane == true)
+		pnmr = PnMR_BM_MD | ((plane->format->pnmr) & ~PnMR_SPIM_MASK) |
+		       PnMR_SPIM_TP_OFF;
+	else
+		pnmr = PnMR_BM_MD | plane->format->pnmr;
 
 	/* Disable color keying when requested. YUV formats have the
 	 * PnMR_SPIM_TP_OFF bit set in their pnmr field, disabling color keying
@@ -625,6 +630,7 @@ int rcar_du_planes_register(struct rcar_du_group *rgrp)
 
 		plane->hwplane = &planes->planes[i + 2];
 		plane->hwplane->zpos = 1;
+		plane->hwplane->fb_plane = false;
 
 		ret = drm_plane_init(rcdu->ddev, &plane->plane, crtcs,
 				     &rcar_du_plane_funcs, plane_formats,
diff --git a/drivers/gpu/drm/rcar-du/rcar_du_plane.h b/drivers/gpu/drm/rcar-du/rcar_du_plane.h
index b2c98b5..140d02d 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_plane.h
+++ b/drivers/gpu/drm/rcar-du/rcar_du_plane.h
@@ -68,6 +68,8 @@ struct rcar_du_plane {
 	unsigned int src_y;
 	unsigned int dst_x;
 	unsigned int dst_y;
+
+	bool fb_plane;
 };
 
 struct rcar_du_planes {
diff --git a/drivers/gpu/drm/rcar-du/rcar_du_regs.h b/drivers/gpu/drm/rcar-du/rcar_du_regs.h
index 2915373..cd76f40 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_regs.h
+++ b/drivers/gpu/drm/rcar-du/rcar_du_regs.h
@@ -321,6 +321,7 @@
 #define PnMR_SPIM_ALP		(1 << 12)	/* Alpha Blending */
 #define PnMR_SPIM_EOR		(2 << 12)	/* EOR */
 #define PnMR_SPIM_TP_OFF	(1 << 14)	/* No Transparent Color */
+#define PnMR_SPIM_MASK		(3 << 12)
 #define PnMR_CPSL_CP1		(0 << 8)	/* Color Palette selected 1 */
 #define PnMR_CPSL_CP2		(1 << 8)	/* Color Palette selected 2 */
 #define PnMR_CPSL_CP3		(2 << 8)	/* Color Palette selected 3 */
-- 
1.8.3.2

